cylc/flow/simulation.py

```diff
@@ -166,46 +220,52 @@ def sim_time_check(
     Returns:
         True if _any_ simulated task state has changed.
     """
-    sim_task_state_changed = False
-    now = time()
+    sim_task_state_changed: bool = False
+    now: float = time()
     for itask in itasks:
         if itask.state.status != TASK_STATUS_RUNNING:
             continue
         # Started time is not set on restart
         if itask.summary['started_time'] is None:
             itask.summary['started_time'] = now
+        if itask.mode_settings is None:
```

Comment:
@MetRonnie

Reply:
I don't think Mypy supports type narrowing for dictionary values.
```diff
diff --git a/cylc/flow/simulation.py b/cylc/flow/simulation.py
index edc693825..b7f74823c 100644
--- a/cylc/flow/simulation.py
+++ b/cylc/flow/simulation.py
@@ -227,18 +227,12 @@ def sim_time_check(
             continue
         # Started time is not set on restart
-        if itask.summary['started_time'] is None:
-            itask.summary['started_time'] = now
+        started_time = itask.summary['started_time']
+        if started_time is None:
+            itask.summary['started_time'] = started_time = now
         if itask.mode_settings is None:
             itask.mode_settings = ModeSettings()
             itask.mode_settings.update(
                 itask, broadcast_mgr)
-        # Ugly hack to make MyPy Happy - It seems quite clear to
-        # me that this is ensured 5-6 lines up ↑
-        if TYPE_CHECKING:
-            assert itask.summary['started_time'] is not None
-        timeout = (
-            itask.summary['started_time'] +
-            itask.mode_settings.simulated_run_length
-        )
+        timeout = started_time + itask.mode_settings.simulated_run_length
         if now > timeout:
             job_d = itask.tokens.duplicate(job=str(itask.submit_num))
```